notebook/notebookapp.py

```diff
@@ -1232,9 +1233,24 @@ def init_server_extensions(self):
             # in the new traitlet
             if not modulename in self.nbserver_extensions:
                 self.nbserver_extensions[modulename] = True
-        
-        for modulename in sorted(self.nbserver_extensions):
-            if self.nbserver_extensions[modulename]:
+
+        # Load server extensions with ConfigManager.
+        # This enables merging on keys, which we want for extension enabling,.
+        # Regular config loading only merges at the class level,
+        # so each level (use > env > system) clobbers the previous.
+        manager = ConfigManager(read_config_path=jupyter_config_path())
```

Comment:
Good call, and further evidence that this is probably too fiddly to be worth doing in-place.